## Prueba cudrados medios
## CHI CUADRADO
### Leandro León 
Realice un programa que permita calcular el valor de Chi-Cuadrada y genere la gráfica de distribución de los 100 primeros números pseudo-aleatorios generados por los métodos de cuadrados medios y productos medios.

Emplee el siguiente nivel de significancia α=0.05

In [40]:
from scipy.stats import chisquare

semilla =74731897457  # input("Escriba semilla: ")
tam=7 #input("Escriba semilla: ")

tam1 = int(tam)
numero1 = int(semilla)
valores =[]
for i in range(100):
    numero2 = (numero1**2)
    snumero2 = str(numero2)
    tam2 = len(snumero2)
    numero=int((tam2/2)-(tam1/2)+1)
    sNumero=str(numero2)
    numUI=sNumero[numero-1:numero+tam1]
    Ui=int(numUI)
    numero1 = int(Ui)
    Rn=Ui/10000
    print(Rn)
    valores.append(Rn/1000)
    


4975.2356
2969.2755
5969.949
291.0626
1743.7118
5308.4145
9264.5038
1030.6602
2604.4786
3087.7785
3760.6506
2492.9352
7259.1139
4734.6131
6561.2066
9432.0478
3525.7014
570.3619
3126.9697
9395.0471
6910.0112
8254.7841
1460.5376
1700.8101
7549.9626
1935.2613
2362.9927
7345.0025
9061.725
4859.9756
9362.8325
2632.423
6508.5092
692.0064
8728.5764
8045.9706
7642.896
3859.2668
3940.2335
5440.0345
3975.3611
3495.8753
1144.1131
9947.8559
9837.0071
6708.6854
6459.7961
8965.6535
2942.682
3773.5312
9537.7173
8051.2947
3346.3462
8032.8902
7324.9652
5115.1812
5078.7088
3283.0751
8582.1122
2649.8133
5105.2485
3562.2467
9601.5516
9793.1274
5344.2726
1249.6231
5578.9205
4353.9453
6839.6753
1158.2094
4490.1424
1378.7722
127.7949
3153.6466
4868.7769
4988.5019
5151.2063
4926.3451
8876.0442
4160.6403
927.7059
6382.3689
4632.7756
2609.7599
8467.3564
6124.4046
8331.7045
7299.8753
8179.3955
2510.7454
8424.6362
4495.1023
5944.6874
9308.2837
4145.4396
4669.4772
4017.3213
8870.4274
4482.2586
642.1572


In [41]:
chisquare(valores, ddof=0.05)

Power_divergenceResult(statistic=144.7327699014051, pvalue=0.001857281374632472)

## Prueba Congruencia lineal
a=74731897457,
b=37747318974, 
Xo=7,
M=19


In [47]:
import time
xn = int(time.time())
a=0.9999999999997491
b=37747318974.00
x0=1234567
valores =[]
for i in range(100):
    xn1 = ((a * xn + x0) % b) / b
    valores.append(xn1)
    print(xn1)
    xn = xn1

print(len(valores))

0.04273160088828076
3.270608552575506e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.270608439457818e-05
3.2706084394

In [45]:
chisquare(valores, ddof=19) # m == ddof == 19

Power_divergenceResult(statistic=19.307877204319965, pvalue=0.9999999999997491)


El valor estadístico es: 19.307877204319965 y el valor de prueba es 0.9999999999997491 por lo que los números siguen una distribución uniforme.